specs/core/0_beacon-chain.md

```diff
@@ -1322,38 +1292,39 @@ def process_justification_and_finalization(state: BeaconState) -> None:
     # Process justifications
     state.previous_justified_epoch = state.current_justified_epoch
     state.previous_justified_root = state.current_justified_root
-    state.justification_bitfield = (state.justification_bitfield << 1) % 2**64
+    state.justification_bitfield = Bitvector[JUSTIFICATION_BITVECTOR_LENGTH](
```

Comment:
I'm not sure using a `Bitvector` for `justification_bitfield` actually aids in readability and am inclined to leave as is.

Just want to gauge the temperature and see if I'm alone in thinking this

Reply:
Technically we can implement the `<<` operator. Or base bitfield off of `int` instead of `list`. I feel like `byte` (or the old `uint64` if we don't care about size) here would be a much simpler solution, and less error prone (we have to test all of this!).

Also, I still like to see bitfields, but for places where we actually can't fit it into a standard type. Which is phase 1+. This would lift the pressure of SSZ changes now in phase-0 freeze, and we can design towards phase-1 needs.
Another worry is whether or not bitfields should be mutable in the spec. A `list` is. An `int` is not. Here we just overwrite it, so I'm inclined to say we don't need mutability. Maybe in phase 1+.

Last thing, we've coercion of container fields, so I think it's fine to remove the `Bitvector[JUSTIFICATION_BITVECTOR_LENGTH](` wrapper, as the container field-setter will force it into that type.

Reply:
And *personally* (willing to help out here, but carefully), I prefer `bytes`/`bytesN`, enhanced with a set of bitfield operations, for phase 1+, and no bitfields in phase0. Simplicity and an idiomatic preference of bytes (good support everywhere) over custom bitfield logic.

Reply:
I can see that the Bitvector is not strictly necessary for the justification_bitfield. I think it could work perfectly well with Vector[boolean, 4]. The Bitvector is not necessary as an optimisation. However, I would still argue that either of those are better than integer notation:
- for people who are not used to bit manipulation, array notation is much clearer
- bit manipulation depends on more conventions (LSB or MSB first, etc.) than array notation; so even a reader familiar with bit manipulation has to do more to understand what is happening
- with vectors, we can make it very clear that we only ever use the last 4 epochs to estimate finalization
I understand bit manipulation well, and so do both of you obviously, but that does not at all mean that it is a natural way to describe things! An array is much more natural object here. If we want to implement an array shift notation, I fully support that, but I think it will not increase readability (as again, the reader would have to understand how the array shift is defined).


> And _personally_ (willing to help out here, but carefully), I prefer `bytes`/`bytesN`, enhanced with a set of bitfield operations, for phase 1+, and no bitfields in phase0. Simplicity and an idiomatic preference of bytes (good support everywhere) over custom bitfield logic.

Now, on this more general point, I do think that the spec benefits from having the Bitlist at phase 0. That is simply because for a reader of the spec, the Bitlist behaves exactly like a List[boolean, N] as far as the interface is concerned. There is no need for "get_bitfield_bit" and "verify_bitfield"-like operations, which should be handled by types rather than by the spec itself. And for the attestation/custody bitfields, we do want the more efficient storage properties. And these are both already in phase 0 (though we could probably change the custody bitfield).

These are my opinions, of course if the consensus sways against bitfields we should stop here and leave phase 0 as it is (that will however likely mean no bitfields ever as we will be baking the attestation logic in now).

Reply:
Convinced :+1: But want bitfields to serialize/hash well (tests! I can help if needed), and hope the readability improves (avoid wrapping with coercion, and prefer no unpacking magic in the spec doc itself).

Reply:
> avoid wrapping with coercion, and prefer no unpacking magic in the spec doc itself

So I replaced that with a slice notation, have a look if you like it better :)

Reply:
I felt that the whole point of adding bitfield classes is to remove the need to do integer manipulation in the user-facing spec code. So definitely oppose keeping that justification bitfield as uint64 for that reason. Bitvector seems fine to me.